runtime/strategies/create-handle-group.js

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2017 Google Inc. All rights reserved.
+// This code may only be used under the BSD style license found at
+// http://polymer.github.io/LICENSE.txt
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+import {Strategy} from '../../strategizer/strategizer.js';
+import Recipe from '../recipe/recipe.js';
+import RecipeWalker from '../recipe/walker.js';
+import Handle from '../recipe/handle.js';
+
+export default class CreateHandleGroup extends Strategy {
+
+  async generate(inputParams) {
+    return Recipe.over(this.getResults(inputParams), new class extends RecipeWalker {
+      onRecipe(recipe) {
+        // Resolve constraints before assuming connections are free.
+        if (recipe.connectionConstraints.length > 0) return;
+
+        const freeConnections = recipe.handleConnections.filter(hc => !hc.handle && !hc.isOptional);
+        let maximalGroup = null;
+        for (let writer of freeConnections.filter(hc => hc.isOutput)) {
+          let compatibleConnections = [writer];
+          let foundSomeReader = false;
+          for (let reader of freeConnections.filter(hc => hc.isInput && hc !== writer)) {
```

Comment:
I think we should also check that the compatible connections are of different particles, eg in:
```
Particle(in Thing thing1, out Thing thing2)
```
`thing1` and `thing2` shouldn't be grouped.